R/osmdata.R

```diff
@@ -0,0 +1,238 @@
+#' bounding box with the OSM tags "place:city" and "boundary:administrative".
+#' The result is filtered by the city name.
+#'
+#' @param city_name A character string with the name of the city
+#' @param bb Bounding box of class `bbox`
+#' @param crs Coordinate reference system as EPSG code
+#' @param multiple A logical indicating if multiple city boundaries should be
+#'                 returned. By default, only the first one is returned.
+#'
+#' @return An sf object with the city boundary
+#' @importFrom rlang .data
+#' @export
+#'
+#' @examples
+#' bb <- get_osm_bb("Bucharest")
+#' crs <- get_utm_zone_epsg_bbox(bb)
+#' get_osm_city_boundary("Bucharest", bb, crs)
+get_osm_city_boundary <- function(city_name, bb, crs, multiple = FALSE) {
+  # Define a helper function to fetch the city boundary
+  fetch_boundary <- function(key, value) {
+    CRiSp::osmdata_as_sf(key, value, bb)$osm_multipolygons |>
+      dplyr::filter(`name:en` == stringr::str_extract(city_name, "^[^,]+") |
+                    name == stringr::str_extract(city_name, "^[^,]+")) |>
+      sf::st_transform(crs) |>
+      sf::st_geometry()
+  }
+
+  # Try to get the city boundary with the "place:city" tag
+  city_boundary <- tryCatch(fetch_boundary("place", "city"),
+                            error = function(e) NULL)
+
+  # If not found, try again with the "boundary:administrative" tag
+  if (is.null(city_boundary)) {
+    city_boundary <- tryCatch(fetch_boundary("boundary", "administrative"),
+                              error = function(e) NULL)
+  }
+
+  # If still not found, throw an error
+  if (is.null(city_boundary)) {
+    stop("No city boundary found. The city name may be incorrect.")
+  }
+
+  if (length(city_boundary) > 1) {
+    if (!multiple) {
+      message("Multiple boundaries were found. Using the first one.")
+      return(city_boundary[1])
+    } else {
+      message("Multiple boundaries were found. Returning all.")
+    }
+  }
+
+  city_boundary
+}
+
+#' Get the river centreline and surface from OpenStreetMap
+#'
+#' @param river_name The name of the river
+#' @param bb Bounding box of class `bbox`
+#' @param crs Coordinate reference system as EPSG code
+#'
+#' @return A list with the river centreline and surface
+#' @export
+#'
+#' @examples
+#' bb <- get_osm_bb("Bucharest")
+#' crs <- get_utm_zone_epsg_bbox(bb)
+#' get_osm_river("Dâmbovița", bb, crs)
+get_osm_river <- function(river_name, bb, crs) {
+  # Get the river centreline
+  river_centerline <- CRiSp::osmdata_as_sf("waterway", "river", bb)
+  river_centerline <- river_centerline$osm_multilines |>
+    dplyr::filter(.data$name == river_name) |>
+    sf::st_transform(crs) |>
+    sf::st_geometry()
+
+  # Get the river surface
+  river_surface <- CRiSp::osmdata_as_sf("natural", "water", bb)
+  river_surface <- dplyr::bind_rows(river_surface$osm_polygons,
+                                    river_surface$osm_multipolygons) |>
+    sf::st_transform(crs) |>
```

Comment:
If we select all geometries here that intersects the river centerline we end up with some weird extra geometries here. Cropping with the bbox prevents this from happening:
```suggestion
    sf::st_crop(bb) |>
    sf::st_transform(crs) |>
```